tests/io/test_pyarrow.py

```diff
@@ -1122,6 +1123,110 @@ def test_projection_concat_files(schema_int: Schema, file_int: str) -> None:
     assert repr(result_table.schema) == "id: int32"
 
 
+def test_projection_partition_inference(tmp_path: str) -> None:
```

Comment:
nit: i think it would be a good idea to test the high-level table `scan` API instead of `ArrowScan`. 

We would need a data file which does not contain the partition field. To achieve that we can do something like this
1. Create a table with no partitioning
2. Write to table to produce data files
3. Add a new column to the table
4. Add the new column a partition
5. Read the table again

By step 4, we should achieve the setup we want

Reply:
In order to implement those steps, we would need to initialize a catalog in `test_pyarrow.py` for updating and committing to that table (as far as I know). Im unsure if initializing a sql catalog (maybe in a `pytest.fixture`) is a good way to go about it?

Reply:
good point, we can either use a local sql catalog (like [this](https://github.com/apache/iceberg-python/blob/1278e8880c4767287dc69208ced20bd444c37228/tests/catalog/test_sql.py#L102-L111)) or the InMemoryCatalog (like [this](https://github.com/apache/iceberg-python/blob/1278e8880c4767287dc69208ced20bd444c37228/tests/catalog/test_base.py#L268-L270))

Reply:
How will the addition of the new partition to the spec be reflected in the manifest file created in step 2, ensuring that the partition struct contains a value during the table scan?

Reply:
Iceberg's partition evolution rules should take care of that. Partition spec is defined at the table level. During the table scan (at step 4), the partition field will be missing from the manifest file and data file which is where the [column project rules apply](https://iceberg.apache.org/spec/#column-projection)

Reply:
> ensuring that the partition struct contains a value during the table scan

good point, [from the spec](https://iceberg.apache.org/spec/#column-projection) 
```
Return the value from partition metadata if an [Identity Transform](https://iceberg.apache.org/spec/#partition-transforms) exists for the field and the partition value is present in the partition struct on data_file object in the manifest.
```

specifically, `the partition value is present in the partition struct on data_file object in the manifest`.

Looks like in order to create this, we need to write data files on an already partitioned table. Otherwise, the data_file's partition value will not be populated.

Reply:
Given a table 
```
CREATE TABLE orders (
  foo BIGINT,
  bar BIGINT,
)
PARTITIONED BY (foo);
USING iceberg;
```
we want
1. data files without the partition field `foo`
2. manifest files referring to data files in step 1 but with partition field value

Reply:
So maybe we can:
1. Create an unpartitioned table and append some data in order to create a data file
2. Create another table with the same schema as table in step 1 + a partition
3. Append manually with `add_files` to table 2 the data file produced by table 1

Will try to implement this tomorrow as I'm afk

Reply:
that seems right to me! Thank you!